File: NBlast.Storage.Core/Env/ConfigReader.fs
namespace NBlast.Storage.Core

open System
open NLog
open System.Configuration


type AppSettingNotFoundException = 
    inherit InvalidOperationException
    new (key) = { inherit InvalidOperationException(sprintf "App setting not found for key %s" key) } 

type AppSettingTryCastException = 
    inherit InvalidOperationException
    new (key, ``type``) = { inherit InvalidOperationException(sprintf "Cannot conver setting %s to %A" key ``type``) } 



type IConfigReader = interface
    abstract member Read: string -> string
    abstract member ReadAsInt: string -> int
end


type ConfigReader() = 
    static let logger = NLog.LogManager.GetCurrentClassLogger()

    interface IConfigReader with 
        member me.Read key =
            try
                ConfigurationManager.AppSettings.[key] |> Environment.ExpandEnvironmentVariables
            with
            | _ -> raise (new AppSettingNotFoundException(key))

        member me.ReadAsInt key =
            match key |> (me :> IConfigReader).Read |> Int32.TryParse with
            | (false, _) -> raise (new AppSettingTryCastException(key, int32.GetType()))
            | (true, value) -> value

File: NBlast.Rest/Configuration/ConfigReader.cs
using System;
using System.Configuration;
using Serilog;
using static LanguageExt.Prelude;

namespace NBlast.Rest.Configuration
{
    public class ConfigReader : IConfigReader
    {
        private static readonly ILogger Logger = Log.Logger.ForContext<ConfigReader>();

        public string Read(string key)
        {
            try
            {
                var appSetting = ConfigurationManager.AppSettings[key];
                return Environment.ExpandEnvironmentVariables(appSetting);
            }
            catch (Exception e)
            {
                Logger.Error(e, $"App setting not found for key {key}");
                throw;
            }
        }

        public int ReadAsInt(string key)
        {
            return parseInt(Read(key)).Match(
                Some: x => x,
                None: () => {
                    var message = $"Cannot conver setting {key} to {typeof(int)}";
                    var exception = new InvalidOperationException(message);
                    Logger.Error(exception, message);
                    throw exception;
                } 
            );
        }
    }
}